File: word4you-python/word_processor.py
from gemini_client import GeminiClient
from utils import prepend_to_vocabulary_notebook, commit_and_push_changes, format_commit_message
from rich.console import Console
from rich.markdown import Markdown
from rich.prompt import Confirm
import re

console = Console()

class WordProcessor:
    """Handles word processing and integration."""
    
    def __init__(self):
        """Initialize the word processor."""
        self.gemini_client = GeminiClient()
    
    def process_word(self, word: str) -> bool:
        """
        Process a word: get explanation, save to wordbook, and commit changes.
        
        Args:
            word: The English word to process
            
        Returns:
            bool: True if successful, False otherwise
        """
        try:
            # Validate word
            if not self._validate_word(word):
                console.print(f"❌ Invalid word: {word}", style="red")
                return False
            
            console.print(f"🔍 Processing word: {word}", style="blue")
            
            # Get explanation from Gemini
            console.print("🤖 Querying Gemini API...", style="yellow")
            explanation = self.gemini_client.get_word_explanation(word)
            
            # Display the explanation
            console.print("\n📖 Word Explanation:", style="green")
            md = Markdown(explanation)
            console.print(md)
            
            # Ask for user confirmation with options
            console.print("\n" + "="*50, style="blue")
            while True:
                console.print("\n[bold]Choose an action:[/bold]")
                console.print("[green]s[/green] - Save to vocabulary notebook")
                console.print("[yellow]r[/yellow] - Regenerate explanation") 
                console.print("[blue]p[/blue] - Preview what will be saved")
                console.print("[red]k[/red] - Skip this word")
                console.print()
                
                action = console.input("Enter your choice (s/r/p/k): ").lower().strip()
                
                if action in ['s', 'save']:
                    # Save to vocabulary notebook
                    console.print("\n💾 Saving to vocabulary notebook...", style="yellow")
                    prepend_to_vocabulary_notebook(explanation)
                    
                    # Commit and push changes
                    console.print("📝 Committing changes...", style="yellow")
                    commit_message = format_commit_message(word)
                    commit_and_push_changes(commit_message)
                    
                    console.print(f"✅ Successfully processed and saved word: {word}", style="green")
                    return True
                    
                elif action in ['r', 'regenerate']:
                    console.print("🔄 Regenerating explanation...", style="yellow")
                    explanation = self.gemini_client.get_word_explanation(word)
                    
                    console.print("\n📖 New Word Explanation:", style="green")
                    md = Markdown(explanation)
                    console.print(md)
                    console.print("\n" + "="*50, style="blue")
                    continue  # Ask again
                    
                elif action in ['p', 'preview']:
                    console.print("\n📋 [bold]Preview of what will be saved:[/bold]", style="blue")
                    console.print("="*50, style="blue")
                    # Show first few lines as preview
                    lines = explanation.split('\n')
                    preview_lines = lines[:10]  # Show first 10 lines
                    console.print('\n'.join(preview_lines))
                    if len(lines) > 10:
                        console.print(f"\n... and {len(lines) - 10} more lines")
                    console.print("="*50, style="blue")
                    continue
                    
                elif action in ['k', 'skip']:
                    console.print("❌ Word explanation skipped.", style="yellow")
                    return True
                    
                else:
                    console.print("❓ Invalid choice. Please enter 's', 'r', 'p', or 'k'.", style="red")
                    continue
            
        except Exception as e:
            console.print(f"❌ Error processing word: {str(e)}", style="red")
            return False
    
    def _validate_word(self, word: str) -> bool:
        """
        Validate that the input is a valid English word.
        
        Args:
            word: The word to validate
            
        Returns:
            bool: True if valid, False otherwise
        """
        if not word or not word.strip():
            return False
        
        # Remove extra whitespace
        word = word.strip()
        
        # Check if it contains only letters and hyphens
        if not re.match(r'^[a-zA-Z\-]+$', word):
            return False
        
        # Check length
        if len(word) < 1 or len(word) > 50:
            return False
        
        return True
    
    def test_api_connection(self) -> bool:
        """Test the API connection."""
        try:
            console.print("🔍 Testing Gemini API connection...", style="yellow")
            success = self.gemini_client.test_connection()
            if success:
                console.print("✅ Gemini API connection successful", style="green")
            else:
                console.print("❌ Gemini API connection failed", style="red")
            return success
        except Exception as e:
            console.print(f"❌ API connection error: {str(e)}", style="red")
            return False 

File: word4you-rust/src/word_processor.rs
use anyhow::Result;
use console::{style, Term};
use dialoguer::Select;
use termimad::*;
use crate::gemini_client::GeminiClient;
use crate::utils::{commit_and_push_changes, format_commit_message, prepend_to_vocabulary_notebook, validate_word};
use crate::config::Config;

pub struct WordProcessor {
    gemini_client: GeminiClient,
    config: Config,
}

impl WordProcessor {
    pub fn new(config: Config) -> Self {
        let gemini_client = GeminiClient::new(config.gemini_api_key.clone());
        Self {
            gemini_client,
            config,
        }
    }

    pub async fn process_word(&self, term: &Term, word: &str, raw: bool) -> Result<()> {
        // Validate word
        validate_word(word)?;
        
        if !raw {
            term.write_line(&format!("🔍 Processing word: {}", word))?;
            term.write_line("🤖 Querying Gemini API...")?;
        }
        
        // Get explanation from Gemini
        let mut explanation = Box::new(self.gemini_client
            .get_word_explanation(word, &self.config.gemini_prompt_template)
            .await?);
        
        // If raw mode, just print the response and return
        if raw {
            println!("{}", explanation);
            return Ok(());
        }
        
        // Display the explanation with beautiful markdown rendering
        term.write_line("\n📖 Word Explanation:")?;
        term.write_line(&style("=".repeat(50)).blue().to_string())?;
        
        // Create a markdown skin for beautiful rendering
        let skin = make_skin();
        
        // Render the markdown
        let text = FmtText::from(&skin, &explanation, None);
        term.write_line(&text.to_string())?;
        
        term.write_line(&style("=".repeat(50)).blue().to_string())?;
        
        // Ask for user confirmation with options
        loop {
            term.write_line("\nChoose an action:")?;
            term.write_line(format!("{} - Save to vocabulary notebook", style("s").green().to_string()).as_str())?;
            term.write_line(format!("{} - Regenerate explanation", style("r").yellow().to_string()).as_str())?;
            term.write_line(format!("{} - Preview what will be saved", style("p").blue().to_string()).as_str())?;
            term.write_line(format!("{} - Skip this word", style("k").red().to_string()).as_str())?;
            term.write_line("")?;
            
            let choices = vec!["s", "r", "p", "k"];
            let selection = Select::new()
                .with_prompt("Enter your choice")
                .items(&choices)
                .default(0)
                .interact()?;
            
            match selection {
                0 => {
                    // Save to vocabulary notebook using the shared method
                    self.save_word(term, word, &explanation)?;
                    return Ok(());
                }
                1 => {
                    // Regenerate explanation
                    term.write_line("🔄 Regenerating explanation...")?;
                    let new_explanation = self.gemini_client
                        .get_word_explanation(word, &self.config.gemini_prompt_template)
                        .await?;
                    explanation = Box::new(new_explanation);
                    
                    term.write_line("\n📖 New Word Explanation:")?;
                    term.write_line(&style("=".repeat(50)).blue().to_string())?;
                    
                    // Render the new markdown
                    let text = FmtText::from(&skin, &explanation, None);
                    term.write_line(&text.to_string())?;
                    
                    term.write_line(&style("=".repeat(50)).blue().to_string())?;
                    continue; // Ask again
                }
                2 => {
                    // Preview
                    term.write_line("\n📋 Preview of what will be saved:")?;
                    term.write_line(&style("=".repeat(50)).blue().to_string())?;
                    
                    let lines: Vec<&str> = explanation.lines().collect();
                    let preview_lines = if lines.len() > 10 {
                        &lines[..10]
                    } else {
                        &lines
                    };
                    
                    for line in preview_lines {
                        term.write_line(line)?;
                    }
                    
                    if lines.len() > 10 {
                        term.write_line(&format!("\n... and {} more lines", lines.len() - 10))?;
                    }
                    
                    term.write_line(&style("=".repeat(50)).blue().to_string())?;
                    continue;
                }
                3 => {
                    // Skip
                    term.write_line("❌ Word explanation skipped.")?;
                    return Ok(());
                }
                _ => {
                    term.write_line("❓ Invalid choice. Please try again.")?;
                    continue;
                }
            }
        }
    }

    pub async fn test_api_connection(&self) -> Result<bool> {
        self.gemini_client.test_connection().await
    }

    pub fn save_word(&self, term: &Term, word: &str, content: &str) -> Result<()> {
        // Validate word
        validate_word(word)?;
        
        term.write_line(&format!("💾 Saving word '{}' to vocabulary notebook...", word))?;
        // Save to vocabulary notebook
        prepend_to_vocabulary_notebook(&self.config.vocabulary_notebook_file, content)?;
        
        // Commit and push changes
        term.write_line("✅ Successfully saved word locally")?;
        term.write_line("📝 Committing changes...")?;
        let commit_message = format_commit_message(word);
        commit_and_push_changes(&commit_message, &self.config.vocabulary_notebook_file, self.config.git_remote_url.as_deref())?;
        
        term.write_line(&format!("✅ Successfully saved word: {}", word))?;
        
        Ok(())
    }
}

fn make_skin() -> MadSkin {
    let mut skin = MadSkin::default();
    
    // Configure colors for different markdown elements
    skin.set_headers_fg(rgb(255, 187, 0));
    skin.bold.set_fg(rgb(255, 187, 0));
    skin.italic.set_fg(rgb(215, 255, 135));
    skin.bullet = StyledChar::from_fg_char(rgb(255, 187, 0), '•');
    skin.quote_mark = StyledChar::from_fg_char(rgb(0, 187, 255), '│');
    skin.quote_mark.set_fg(rgb(0, 187, 255));
    skin.inline_code.set_fg(rgb(255, 119, 119));
    skin.inline_code.set_bg(rgb(40, 44, 52));
    skin.code_block.set_bg(rgb(40, 44, 52));
    skin.code_block.set_fg(rgb(255, 119, 119));
    
    skin
} 